### DISCUSSION TITLE
The snippet cmp is falsely triggered by a symbol

### DISCUSSION THREAD

**Discussion by winkee01** [hrsh7th/nvim-cmp#954] [Q&A]
Usually, snippet cmp was triggered by several prefix words, but If a snippet contains a symbol (eg. `,` or `;` etc), the snippet cmp was triggered as long as I type the symbol, for example

<img width="292" alt="CleanShot 2022-05-02 at 21 59 05@2x" src="https://user-images.githubusercontent.com/49930785/166246896-56e0cc04-7b82-436f-b9d2-bbd4a9efd250.png">

the symbol is in the middle of the snippet definition, but it was triggered immediately after I type it, this is very annoying.

The expected behavior should be that only when I typed the whole snippet keyword should it be triggered.

the snippet definition is

```
snippet pri, "print! with format param"
	print!("${1}{${2}}", ${3});

snippet pln, "println! with format param"
	println!("${1}{${2}}", ${3});
```

**Comment by hrsh7th** [comment]
It is not falsly. For example, if we define `/**` snippet, it should be shown via symbol.

**Reply by winkee01** [reply]
But it shouldn't be shown when the symbol is in the middle and I didn't even type the whole prefix. It should be only triggered when I type `prl,`, not when I type a single `,`

**Reply by hrsh7th** [reply]
You can configure matching behaivior. See docs

**Reply by hrsh7th** [reply]
Whether or not the trigger keyword string starts with a symbol can only be known by performing completion and starting the matching process.
In other words, it is a contradiction that you say ", is not the beginning of the keyword string and should not be completed".

_google translated_

**Reply by winkee01** [reply]
I don't know what you are justifying, I type `,` in an empty line, and it triggered the completion, this is absolutely wrong.

**Reply by hrsh7th** [reply]
I say again if you define `/**` snippet. I think nvim-cmp should trigger completion via press `/`.

So nvim-cmp should trigger via symbol.

1. nvim-cmp should invoke completion via `/`
2. nvim-cmp should do fuzzy matching
3. If you type `,`, nvim-cmp will invoke completion (because 1
4. if you type `,`, nvim-cmp will show `pri,~1 (because 2

**Reply by hrsh7th** [reply]
I've received a bug report that # doesn't show `#!/usr/bin`.

>  this is absolutely wrong.

Please consider more wide usecase.

**Reply by winkee01** [reply]
I don't know what you are talking about! `/**` can be triggered by `/` is ok because `/` is the the leading char, `pri,` should be triggered by `pri`, not **`,`**, **`,`** is not the leading char.

**Reply by winkee01** [reply]
> 2\. nvim-cmp should do fuzzy matching

to fuzzy matching a single `,` from `pri,` is wrong and should be considered a bug, this is annoying.

**Reply by hrsh7th** [reply]
It is only after performing the completion process that nvim-cmp can know that "there are items that start with / or,".

Therefore, it is not possible to decide whether to trigger with `/` by whether the user has defined a snippet of `/** `.

**Reply by winkee01** [reply]
I am trying to understand why you do a fuzzy matching over prefix matching, don't make too much sense to me. And I think this should be a bug if I type a single `,` and it gives me a nonsense completion candidate.

**Reply by hrsh7th** [reply]
There are two ways I can think of to achieve this behavior.

1. Add special processing to the matching processing of completion candidates
2. Completion source side filters candidates in advance

**Reply by hrsh7th** [reply]
BTW, Do you read the docs?
https://github.com/hrsh7th/nvim-cmp/blob/main/doc/cmp.txt#L433

**Reply by winkee01** [reply]
I remembered that when I use UltiSnip engine in the old days, it can be configured to not trigger the snippet when the cursor is not at the start of a line, so based on that understanding, a cmp engine should be able to handle different cases, especially this one.

**Reply by hrsh7th** [reply]
If you think it's possible with rational solution, feel free to send PR.
Sorry. I can't think the rational solution about it.

**Reply by winkee01** [reply]
Under what situation do you think we can't filter the cmp candidate out when I type `,` and it gives me `pri,`, at least I can't think of a use case that this can't be filtered out.

**Reply by hrsh7th** [reply]
Do you insist that "the case where the input string starts with a symbol should be treated specially"?
Can it be expressed as a clear specification?

https://github.com/hrsh7th/nvim-cmp/blob/main/lua/cmp/matcher.lua#L15

**Reply by hrsh7th** [reply]
You seem to suggest that I should implement the following:

> 1. Add special processing to the matching processing of completion candidates

**Reply by winkee01** [reply]
Probably, from my point of view, it's probably rational, I will check it later to see if I missed something, but at the time, I think it makes sense.

**Reply by winkee01** [reply]
I don't quite understand how you do the matching because say I have `pri,` defined to be a snippet, when I type `,`, it gives me `pri,`, but when I type `ri`, it gives me nothing. Shouldn't it also give me `pri,` ? because it does fuzzy matching, and once it found something, cmp is triggered?  it seems you already do some special processing to symbols.

**Reply by hrsh7th** [reply]
I also think about rational specifications, but I haven't got an image so far.
Wait for a suggestion.

**Reply by winkee01** [reply]
the even more annoying things is that even if I type `,,,` or `xx,`, it still triggers the cmp, so basically, as long as it contains a symble, it triggers, no matter where the symbol appears in your input.
<img width="304" alt="CleanShot 2022-05-03 at 01 28 14@2x" src="https://user-images.githubusercontent.com/49930785/166295570-cfb8870b-88c3-435d-b47e-93aaf9bc926a.png">
<img width="304" alt="CleanShot 2022-05-03 at 01 30 13@2x" src="https://user-images.githubusercontent.com/49930785/166295706-99625353-0939-4379-a6be-467f6b624731.png">

**Reply by hrsh7th** [reply]
It's depending on the source implementation.

**Reply by winkee01** [reply]
today, I reviewed your code, and trying to debug, how did you get `e:get_word()` and `e:get_filter_text()`? when I type `,`, they both are `pri,`, so it might be the problem.

**Reply by Shougo** [reply]
Hm.  The filter seems doesn't work.
Please create minimal init.vim.
Other people can reproduce the problem.

**Reply by Shougo** [reply]
It seems the comma matches fuzzy.
And if the candidate is displayed, it can be expanded by nvim-cmp auto snippet expansion.

**Reply by hrsh7th** [reply]
See this commit. https://github.com/hrsh7th/nvim-cmp/pull/956/commits/766e17a23e12d556db04a03a5ec67f91bdd31036

This is what it means to "define a specification".

**Reply by winkee01** [reply]
I have been thinking of a more general solution for this almost the whole day 😂, I guess a more general fuzzy strategy(I guess the same meaning to your spec) might be better? instead of taking a special process to symbols, why not make a strategy, that can apply to the filter. A strategy is saying users can define that he doesn't want to trigger the cmp when some conditions are not meet. For example, a symbol can be a strategy, or the matcher can simply ignore some input patterns.

**Reply by winkee01** [reply]
for example, on the user side, he can define something like this:
```
cmp.setup({
   snippet = {
     expand = function(args) require('luasnip').lsp_expand(args.body) end
     ignore_pattern = function(input) return input[0] = ',' end -- strategy, if input is a single comma, don't do fuzzy
   }
   ignore_pattern = function(input, pos)   -- strategy: if input is less than two chars, or, it's not at the line beginning, don't do fuzzy
     if (#input < 2) or pos > 0 then
       return true
     end
   end 
}
```
A strategy engine might be a bit complex, but I think a simpler one might be useful enough. The basic idea is just to make it a bit smarter.
